### PR TITLE
Added modules for building VB6 projects with SxS manifests

### PR DESCRIPTION
**Pull request by Vidarls**
Support for buiilding VB6 projects and embedding SxS manifests for registration free interop deployments.

Tested on our inhouse legacy codebase with great success so far.

Due to our codebase having a large number of projects and dependencies, I've tried to paralellize as much as possible. Result: Current powershell build: approx 4 minutes, now with FAKE: ~30 sec.

Contains loads of hacs and cheats, feel free to improve and suggest changes :smile:

### CODE REVIEW

File: src/app/FakeLib/Sxshelper.fs [hunk 1]
```diff
@@ -0,0 +1,354 @@
+﻿/// Module that enables creating and embedding Side-by-Side interop
+/// manifests for registration free deployment of Com-.net interop projects
+module Fake.Sxshelper
```
Comment (forki) [inline src/app/FakeLib/Sxshelper.fs]: Helper is always upper-case in our lib. How about SxSHelper?

File: src/app/FakeLib/Sxshelper.fs [hunk 2]
```diff
@@ -0,0 +1,354 @@
...
+
+open Fake
+open System
+open System.IO
+open System.Linq
+open System.Xml.Linq
+
+/// Represents a `.net` assembly that may be used in COM interop projects`
```
Comment (forki) [inline src/app/FakeLib/Sxshelper.fs]: `.NET`?

Comment (forki) [inline src/app/FakeLib/Sxshelper.fs]: also there is an additional backtick at the end of the line

File: src/app/FakeLib/Sxshelper.fs [hunk 3]
```diff
@@ -0,0 +1,354 @@
...
+/// Represents an executable to create an _application manifest_ for
+type InteropApplicationData = 
+    {
+        /// Path of the executable binary file
+        ExecutablePath:string
+        
+        /// Version of the executable
+        Version:String
+
+        /// Depandant `.net`assemblies of the executable
```
Comment (forki) [inline src/app/FakeLib/Sxshelper.fs]: Dependent `.NET` assemblies of the executable <- small typos

File: src/app/FakeLib/Sxshelper.fs [hunk 4]
```diff
@@ -0,0 +1,354 @@
...
+type private ILparsingResult =
+    /// Found all required data
+    /// Includes structured assembly data
+    | Success of InteropAssemblyData
+    /// Failed to find all reguired data
+    /// Includes an error message
+    | Failed of string
+
+let private sdkBasePath = ProgramFilesX86 @@ "Microsoft SDKs\Windows"
+let private netFxDir = (Environment.GetEnvironmentVariable("WINDIR")) @@ "Microsoft.NET\Framework"
```
Comment (forki) [inline src/app/FakeLib/Sxshelper.fs]: in EnvironmentHelper we have getInstalledDotNetFrameworks. Does this help?

Comment (forki) [inline src/app/FakeLib/Sxshelper.fs]: if not. can you move this two properties over there?

File: src/app/FakeLib/Sxshelper.fs [hunk 5]
```diff
@@ -0,0 +1,354 @@
...
+
+/// Path to `mt.exe`
+/// ref: https://msdn.microsoft.com/en-us/library/aa375649(v=vs.85).aspx
+let private  mtToolPath = !! (sdkBasePath + "/**/mt.exe") -- (sdkBasePath + "/**/x64/*.*") 
+                          |> getNewestTool
+
+/// Path to `regasm.exe`
+/// Reason for this to be public is that the VB6 helpe uses this
+/// but the `getNewestToll` function is in this module
+let regAsmToolPath = !! (netFxDir + "/**/RegAsm.exe")
```
Comment (forki) [inline src/app/FakeLib/Sxshelper.fs]: we have a regasmHelper. Your improvements should be applied there.

File: src/app/FakeLib/Sxshelper.fs [hunk 6]
```diff
@@ -0,0 +1,354 @@
...
+/// Path to `regasm.exe`
+/// Reason for this to be public is that the VB6 helpe uses this
+/// but the `getNewestToll` function is in this module
+let regAsmToolPath = !! (netFxDir + "/**/RegAsm.exe")  
+                             |> getNewestTool
+
+/// Path to `ildasm.exe
+/// .net fx dissasembly tool
+/// ref: https://msdn.microsoft.com/en-us/library/f7dy01k1(v=vs.110).aspx
+let private ildasmPath = !! (sdkBasePath + "/**/ildasm.exe") -- (sdkBasePath + "/**/x64/*.*")
```
Comment (forki) [inline src/app/FakeLib/Sxshelper.fs]: very  cool. Can you create a Ildasm helper by copying the regasm helper?

File: src/app/FakeLib/Vb6helper.fs [hunk 7]
```diff
@@ -0,0 +1,307 @@
+﻿/// Enables building of Visual Basic 6 projects
+/// Also includes a do-it-all function that will embed interop
+/// side-by-side manifest to executables from Vb6 using
+/// functions from the Side-by-side helper module
+module Fake.Vb6helper
```
Comment (forki) [inline src/app/FakeLib/Vb6helper.fs]: same here. please make it `VB6Helper`

File: src/app/FakeLib/Vb6helper.fs [hunk 8]
```diff
@@ -0,0 +1,307 @@
...
+/// This function will:
+///
+/// 1. Determine which of the `possibleAssemnblies` are referenced by any of the provided `.vbp` files
+/// 2. Temporarily register any referenced assemblies using `RegAsm /codebase /tlb`
+/// 3. Run VB6 command line make on all provided `.vbp` projects
+/// 4. Unregister all registered assemblies
+/// 5. Generate and embed Side-By-Side interop appplication manifests in all generated VB6 executables
+/// 6. Generate and embed Side-By-Side interop assembly manifest in all referenced assemblies 
+let public BuildAndEmbedInteropManifests (getConfig: Vb6BuildParams->Vb6BuildParams) (vb6Projects: string seq) (possibleAssemblies: string seq) =
+    let config = defaultVb6BuildParams |> getConfig
```
Comment (forki) [inline src/app/FakeLib/Vb6helper.fs]: For tasks please add trace start/end functions like in

https://github.com/fsharp/FAKE/blob/master/src/app/FakeLib/FixieHelper.fs#L38
https://github.com/fsharp/FAKE/blob/master/src/app/FakeLib/FixieHelper.fs#L59

**Comment by forki** [comment]
this is very cool stuff.

File: src/app/FakeLib/Sxshelper.fs (hunk 6 above)
Comment (Vidarls) [inline src/app/FakeLib/Sxshelper.fs]: Might be possible, but there are so many options for the ildasm tool that I wouldn't know how to create something that is reusable for anything useful. Currently I know only this specific usecase. Might be better to wait, and consider adding a generic helper if more usecases would occur.

(To be frank, I would very much like to find some other way of getting the metadata, that does not rely on the porcelain output of ildasm)

**Comment by Vidarls** [comment]
Also more interresting travis failures for our collection:

```
$ ./build.sh Default
No version specified. Downloading latest stable.
The remote server returned an error: (403) Forbidden.
```

File: src/app/FakeLib/Sxshelper.fs (hunk 6 above)
Comment (forki) [inline src/app/FakeLib/Sxshelper.fs]: fair enough. +1

**Comment by forki** [comment]
yep. I really need to mirror paket somewhere elese.

**Comment by Vidarls** [comment]
I believe I've addressed most of your concerns now.

**Comment by forki** [comment]
I'm on vacation this week, but I will review this when

**Comment by Vidarls** [comment]
Ok, hope you have a good vacation :sunny:

**Comment by forki** [comment]
very very cool. thanks.

**Comment by Vidarls** [comment]
Thank you :smile:

**Pull request merged**